### ISSUE TITLE
Endless loop in BPROG (rat(1/(x^(2/3)+1)), algebraic)

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2667]
Imported from SourceForge on 2024-07-06 17:14:23
Created by **[tomasriker](https://sourceforge.net/u/tomasriker/)** on 2018-04-11 13:55:43
Original: https://sourceforge.net/p/maxima/bugs/3419

---

These loop forever:

~~~
rat(1/(x^(2/3)+1)), algebraic;
rat(1/(x^(2/5)+1)), algebraic;
rat(1/(x^(2/7)+1)), algebraic;
...
~~~

An endless loop takes place in the function `BPROG` in `SIMP.LISP`.

Unfortunately, that function is completely undocumented and its name isn't very descriptive.

`BPROG` has a loop (using a `GO` form) and a conditional return. During the evaluation of `rat(1/(x^(2/3)+1)), algebraic`, that condition is never met, so the function loops forever.

It seems easy to detect that particular case of an endless loop, though, by checking whether a certain variable named `P2B` changes from one iteration to the next or not. My theory is that if it remaings unchanged, then the function will loop forever. And so I made a little fix by checking for that condition and instead of looping forever, returning what would be returned if the exit condition was met.  For my particular case, the result seems OK, and the change didn't break any tests.

Take this line:
~~~
(setq p1b (pquotientchk (psimp var (caddr f1)) a))
~~~

Replace it by:
~~~
(when (equal p1b (setq p1b (pquotientchk (psimp var (caddr f1)) a)))
  (return (cons (ratreduce (ptimes (cdr r) coef2r) p2b)
    (ratreduce (ptimes (cdr s) coef2s) p2b))))
~~~

Of course, it would be much better to actually fix the cause of the endless loop, but that would require some understanding of `BPROG` and its purpose.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:14:25
Created by **[tomasriker](https://sourceforge.net/u/tomasriker/)** on 2018-04-11 14:47:59
Original: https://sourceforge.net/p/maxima/bugs/3419/#f97c

---

That fix was not enough. It doesn't cover this case:
~~~
rat(1/((x-1)^(2/3)+1)), algebraic;
~~~
Here, the variable `P1B` changes with every iteration, but enters a cycle of length 2.
My new "fix" keeps all values seen for `P1B` in a list. If it sees one that it has seen before, it causes the function to exit.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:14:29
Created by **[tomasriker](https://sourceforge.net/u/tomasriker/)** on 2018-04-19 10:58:15
Original: https://sourceforge.net/p/maxima/bugs/3419/#5327

---

- Description has changed:

Diff:

~~~~

--- old
+++ new
@@ -13,7 +13,7 @@
 
 `BPROG` has a loop (using a `GO` form) and a conditional return. During the evaluation of `rat(1/(x^(2/3)+1)), algebraic`, that condition is never met, so the function loops forever.
 
-It seems easy to detect that particular case of an endless loop, though, by checking whether a certain variable named `P1B` changes from one iteration to the next or not. My theory is that if it remaings unchanged, then the function will loop forever. And so I made a little fix by checking for that condition and instead of looping forever, returning what would be returned if the exit condition was met.  For my particular case, the result seems OK, and the change didn't break any tests.
+It seems easy to detect that particular case of an endless loop, though, by checking whether a certain variable named `P2B` changes from one iteration to the next or not. My theory is that if it remaings unchanged, then the function will loop forever. And so I made a little fix by checking for that condition and instead of looping forever, returning what would be returned if the exit condition was met.  For my particular case, the result seems OK, and the change didn't break any tests.
 
 Take this line:
 ~~~

~~~~

- **status**: open --> closed
